Initial implementation of HVAC zone control and diagnostics

bitfliq opened the pull request:
Updated description: displays zone temperature, humidity, CFM, and percent open. Temperature setpoints are controllable.

barndawgie commented on econet_hvac_zone_template.yaml:
```diff
@@ -0,0 +1,60 @@
+
+binary_sensor:
+  - platform: econet
+    name: ZON${zoneid}ENAB
```
This name will show up in the front end so it should be somehting more readable.

bitfliq replied:
fixed

bitfliq replied:
removed this sensor altogether.

barndawgie commented on example-local.yaml:
```diff
@@ -3,10 +3,37 @@ packages:
     vars:
       external_components_source: components
+  
+  # for zones - include the econet hvac package above and the appropriate zones below
+  # because we are re-using esphome-econet for each inclusion, the path is required also.
+  # zone1: !include 
+  #    url: https://github.com/esphome-econet/esphome-econet
+  #   file: econet_hvac_zone_template.yaml
+  #   vars:
+  #     zoneid: "1"
+  #     zonename: "Zone1"
+  #     requestmod: 5
+  # zone2: !include 
+  #    url: https://github.com/esphome-econet/esphome-econet
+  #   file: econet_hvac_zone_template.yaml
+  #   vars:
+  #     zoneid: "2"
+  #     zonename: "Zone2"
+  #     requestmod: 6
+  # zone3: !include 
+  #    url: https://github.com/esphome-econet/esphome-econet
+  #   file: econet_hvac_zone_template.yaml
+  #   vars:
+  #     zoneid: "3"
+  #     zonename: "Zone3"
+  #     requestmod: 7
```
This implies users will have to manually configure each zone, correct? Vs. us somehow detecting what zones are present and dynamically creating the right sensors?

bitfliq replied:
For single zone installs, user would want to include econet_hvac (line 7) in this file, and that file includes _base.  For multi-zone, econet_hvac_base and all the appropriate zones.  The zone files do not include the _base file due to repeats.

bitfliq replied:
Correct.  Dynamically creating the zones would be a far more elegant approach, but not quite sure how to go about that.  Wondering if a reasonable compromise here would be to query if the zone is enabled and surface that to the user in HA?

bitfliq replied:
This is refactored with current code.

barndawgie commented on econet_hvac_base.yaml:
What's the reason that we need to extract these to a base file? It doesn't seem like it's being used/referenced in more than one place, or am I missing something?

bitfliq replied:
(re-pasted from wrong thread above)
For single zone installs, user would want to include econet_hvac (line 7) in this file, and that file includes _base. For multi-zone, econet_hvac_base and all the appropriate zones. The zone files do not include the _base file due to repeats.

barndawgie commented:
I added a few comments - maybe just for my clarification as I haven't fully been following the threads in Discord on this.

Also, please check the pre-commit ci failure. You can validate that stuff locally by [installing pre-commit](https://pre-commit.com/):
```bash
pip install pre-commit
pre-commit install  # Run from base directory of the project to automatically check before commiting
pre-commit run --all-files  # You can run this to check every file in your local branch
```

tronikos commented on econet_hvac_zone_template.yaml:
```diff
@@ -0,0 +1,56 @@
+    entity_category: "diagnostic"
+    disabled_by_default: true
+    request_mod: ${requestmod}
+    request_once: true
```
Is this intentional?

tronikos commented on example-local.yaml:
```diff
@@ -3,14 +3,41 @@ packages:
+  #   vars:
+  #     zoneid: "2"
+  #     zonename: "Zone2"
+  #     requestmod: 6
```
Do you really need to increment request_mod?

bitfliq replied:
different request mods are required in new code due to different src addresses of zone thermostats

tronikos commented on example-local.yaml:
```diff
@@ -3,14 +3,41 @@ packages:
+  #   vars:
+  #     zoneid: "1"
+  #     zonename: "Zone1"
+  #     requestmod: 5
```
You can start with 4. HVAC is using 3 as the largest value.

tronikos commented on example-local.yaml:
```diff
@@ -3,14 +3,41 @@ packages:
     vars:
       external_components_source: components
 
+  # for zones - include the econet hvac package above and the appropriate zones below
```
I think it might be better to revert changes to this file and create a new one: example-hvac-zones.yaml with 3 zones and a comment to adjust (add or remove). It will be easier for users.

bitfliq replied:
agreed

stockmopar commented:
Is there anything we need to change with this to get this pulled in?

bitfliq commented:
Feedback for different implementation per tronikos is main holdup here.

barndawgie commented:
As currently implemented, I think this would require support for `vars:` on remote packages. See https://github.com/esphome/feature-requests/issues/2542

bitfliq commented:
Still working on this.

bitfliq commented:
Ready for feedback.  As mentioned in the discord, the two outstanding issues with this are:

1. users need to copy in or reference a large number of yaml files in their includes (as shown in example-local) to use this.  could be solved by putting the necessary econet_* files into a subdirectory to prevent them from appearing in the esphome dashboard
2. users cannot currently with esphome use variables on remote files

bitfliq commented:
This is ready for review.  Until variables are supported by esphome for remote packages, it will require placing yaml files in the esphome directory (or in a subdirectory).

stockmopar commented:
I think this is 95% of the way there. Should we put a few of these config files into a subdirectory to make it clearer which are breaking if we update/change?

bitfliq commented:
Moved files to a subfolder and refactored how variables are handled. Added an example for using this as a remote package directly from GitHub without manually downloading multiple yaml files. Ready for review.

barndawgie reviewed (approved):
This seems in a good enough state to get into the codebase and we can iterate on the layout of these configuration files.
